content/unroll-default-arguments.md

```diff
@@ -0,0 +1,817 @@
+   though they are _almost_ binary compatible in Scala 2. Direct calls to `unapply` are binary 
+   incompatible, but most common pattern matching of `case class`es goes through a different 
+   code path that _is_ binary compatible. In practice this should be sufficient for 99% of use 
+   cases, but it does mean that it is possible for code written as below to fail in Scala 2
+   if a new unrolled parameter is added to the case class and `.unapply` is called directly.
+
+4. While `@unroll`ed `case class`es are fully binary compatible, they are *not* fully
+   _source_ compatible, due to the fact that pattern matching requires all arguments to
+   be specified. This proposal does not change that. Future improvements related to
+   [Pattern Matching on Named Fields](https://github.com/scala/improvement-proposals/pull/44)
+   may bring improvements here. But as we discussed earlier, binary compatibility is generally
+   more important than source compatibility, and so we do not need to wait for any source
+   compatibility improvements to land before proceeding with these binary compatibility
+   improvements.
+
+5. This proposal does not address how macros will derive typeclasses for `case class`es, and
+   whether or not those will be binary/source/semantically compatible. That is up to the
+   individual macro implementations to decide. e.g., [uPickle](https://github.com/com-lihaoyi/upickle)
+   has a very similar rule about adding `case class` fields, except that field ordering
+   does not matter. Trying to standardize this across all possible macros and all possible
+   typeclasses is out of scope
+
+6. `@unroll` generates a quadratic amount of generated bytecode as more default parameters
+   are added: each forwarder has `O(num-params)` size, and there are `O(num-default-params)`
+   forwarders. We do not expect this to be a problem in practice, as the small size of the
+   generated forwarder methods means the constant factor is small, but one could imagine
+   the `O(n^2)` asymptotic complexity becoming a problem if a method accumulates hundreds of
+   default parameters over time. In such extreme scenarios, some kind of builder pattern
+   (such as those listed in [Major Alternatives](#major-alternatives)) may be preferable.
+
+7. `@unroll` only supports `final` methods. `object` methods and constructors are naturally
+   final, but `class` or `trait` methods that are `@unroll`ed need to be explicitly marked `final`.
+   It has proved difficult to implement the semantics of `@unroll` in the presence of downstream 
+   overrides, `super`, etc. where the downstream overrides can be compiled against by different 
+   versions of the upstream code. If we can come up with some implementation that works, we can
+   lift this restriction later, but for now I have not managed to do so and so this restriction
+   stays.
+
+### Challenges of Non-Final Methods and Overriding
```

Comment:
I'd like to highlight this section as worth discussing.

I would like to be able to support non-final methods, abstract-methods-implemented-by-final-methods, and abstract-methods-implemented-by-non-final-methods. But I haven't managed to come up with an implementation strategy that I can convince myself is correct.

Right now, I'm reasonably confident that:

- `@unroll`ing final methods works
- `@unroll`ing override-able concrete methods cannot be done safely

There's one case I'm not sure about

- `@unroll`ing abstract methods that are only implemented by final methods (i.e. implemented with no subsequent overrides)

If we could make this case work, it would open up the use of `@unroll` in traits and abstract classes, which would be a big win over the status quo of only allowing final concrete methods